(AB-1967391) Draft Getting Started for v2 (#74)
MicrosoftDocs/PowerShell-Docs-DSC: dsc/docs-conceptual/dsc-2.0/getting-started/invoking-dsc-resources.md

---
ms.date: 07/06/2022
keywords:  dsc,powershell,configuration,setup
title:  Get started with invoking DSC Resources
description: >
  This topic explains how to get started using PowerShell Desired State Configuration (DSC) with the
  Invoke-DscResource cmdlet.
---

# Get started with invoking DSC resources

This topic explains how to get started using PowerShell Desired State Configuration (DSC) with the
`Invoke-DscResource` cmdlet.

Starting with version 2, DSC is not supported for Windows PowerShell or applying configuration
documents directly. Instead, you can use the [Invoke-DscResource][cmdlet-Invoke-DscResource] cmdlet
to directly call the methods of a DSC resource to inspect or enforce the state of resources on a
system.

## Installing DSC

> To install version 2 of the PSDesiredStateConfiguration module from the PowerShell Gallery:

Reviewer: ```suggestion
To install version 2 of the **PSDesiredStateConfiguration** module from the PowerShell Gallery:
```

```powershell
Install-Module -Name PSDesiredStateConfiguration -Repository PSGallery -MaximumVersion 2.99
```

> [!IMPORTANT]
> Be sure to include the parameter **MaximumVersion** or you could install version 3 (or higher) of
> **PSDesireStateConfiguration** instead.

## Install a module containing DSC resources

Starting with version 2, DSC does not include any DSC resources. You can install modules with DSC
resources from external sources, such as the [PowerShell Gallery][psgallery].

You can search for DSC resources with the `Find-DscResource` cmdlet. To see the list of modules with
DSC resources, use the following command:

```powershell
Find-DscResource -Repository PSGallery | Select-Object -Property ModuleName, Version -Unique
```

For more information on searching for DSC resources in a PowerShell repository, see
[Find-DscResource][cmdlet-Find-DscResource].

To install the latest version of the PSDscResources module, use the following command:

```PowerShell
Install-Module 'PSDscResources' -Repository PSGallery
```

## Invoking DSC resources

> When calling the `Invoke-DscResource` cmdlet, you specify which method or function to call by using
> the **Method** parameter. You specify the properties of the resource by passing a hashtable as the
> value of the **Property** parameter.

Reviewer: ```suggestion
When calling the `Invoke-DscResource`, you specify which method or function to call using
the **Method** parameter. Specify the properties of the resource as a hashtable value for the **Property** parameter.
```

The following commands show how you can use the resource methods of DSC Resources. They use the
**Environment** DSC resource in the **PSDscResources** module.

### Test whether a resource is correctly configured

The `Test` method inspects the state of a resource and returns an **InvokeDscResourceTestResult**

> object with an **InDesiredState** property of either `$true` (if every specified property is in the
> desired state) or `$false`.

Reviewer: ```suggestion
object with an **InDesiredState** property set to `$true` or `$false` depending on whether every specified property is in the
desired state.
```

```powershell
Invoke-DscResource -Name Environment -Module PSDscResources -Method Test -Property @{
    Name   = 'DSC_EXAMPLE'
    Ensure = 'Present'
    Value  = 'Desired State Configuration'
    Target = 'Process'
}
```

### Ensure an environment variable exists with the correct value

The `Set` method enforces the state of a resource on the system.

```powershell
Invoke-DscResource -Name Environment -Module PSDscResources -Method Set -Property @{
    Name   = 'DSC_EXAMPLE'
    Ensure = 'Present'
    Value  = 'Desired State Configuration'
    Target = 'Process'
}
```

> [!CAUTION]
> Calling `Invoke-DscResource` with `Set` does not verify whether the resource is already in the

> > desired state. To ensure you are not unnecessarily enforcing state, always use the `Test` method
> > to see if invoking with the `Set` method is required.

Reviewer: ```suggestion
> desired state. To ensure you are not unnecessarily enforcing state, always use the `Test` method.
```
